src/renderer/atlas/AtlasEngine.h

```diff
@@ -0,0 +1,748 @@
+        // You can generally assume that only u32 or types of that alignment are allowed.
+        struct Cell
+        {
+            alignas(u32) u16x2 tileIndex;
+            alignas(u32) CellFlags flags;
+            u32x2 color;
+        };
+
+        struct AtlasKeyAttributes
+        {
+            u16 inlined : 1;
+            u16 bold : 1;
+            u16 italic : 1;
+            u16 cellCount : 13;
+
+            ATLAS_POD_OPS(AtlasKeyAttributes)
+        };
+
+        struct AtlasKeyData
+        {
+            AtlasKeyAttributes attributes;
+            u16 charCount;
+            wchar_t chars[14];
+        };
+
+        struct AtlasKey
+        {
+            AtlasKey(AtlasKeyAttributes attributes, u16 charCount, const wchar_t* chars)
+            {
+                const auto size = dataSize(charCount);
+                const auto data = _data.initialize(size);
+                attributes.inlined = _data.would_inline(size);
+                data->attributes = attributes;
+                data->charCount = charCount;
+                memcpy(&data->chars[0], chars, static_cast<size_t>(charCount) * sizeof(AtlasKeyData::chars[0]));
+            }
+
+            const AtlasKeyData* data() const noexcept
+            {
+                return _data.data();
+            }
+
+            size_t hash() const noexcept
+            {
+                const auto d = data();
+#pragma warning(suppress : 26490) // Don't use reinterpret_cast (type.1).
+                return std::_Fnv1a_append_bytes(std::_FNV_offset_basis, reinterpret_cast<const u8*>(d), dataSize(d->charCount));
+            }
+
+            bool operator==(const AtlasKey& rhs) const noexcept
+            {
+                const auto a = data();
+                const auto b = rhs.data();
+                return a->charCount == b->charCount && memcmp(a, b, dataSize(a->charCount)) == 0;
+            }
+
+        private:
+            SmallObjectOptimizer<AtlasKeyData> _data;
+
+            static constexpr size_t dataSize(u16 charCount) noexcept
+            {
+                // This returns the actual byte size of a AtlasKeyData struct for the given charCount.
+                // The `wchar_t chars[2]` is only a buffer for the inlined variant after
+                // all and the actual charCount can be smaller or larger. Due to this we
+                // remove the size of the `chars` array and add it's true length on top.
+                return sizeof(AtlasKeyData) - sizeof(AtlasKeyData::chars) + static_cast<size_t>(charCount) * sizeof(AtlasKeyData::chars[0]);
+            }
+        };
+
+        struct AtlasKeyHasher
+        {
+            size_t operator()(const AtlasKey& key) const noexcept
+            {
+                return key.hash();
+            }
+        };
+
+        struct AtlasValueData
+        {
+            CellFlags flags = CellFlags::None;
+            u16x2 coords[7];

```

Comment:
Yeah AtlasKey and AtlasValue are both exactly 32 bytes large. There's no reason for this except to demonstrate that this works. This will be rewritten to use a custom hash-table in the future were this "inlineability" will be important.